extension/runtime/src/main/java/io/quarkiverse/infinispan/embedded/runtime/cache/InfinispanCacheImpl.java

```diff
@@ -0,0 +1,220 @@
+package io.quarkiverse.infinispan.embedded.runtime.cache;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.util.NullValue;
+import org.infinispan.commons.util.concurrent.CompletionStages;
+
+import io.quarkiverse.infinispan.embedded.runtime.InfinispanEmbeddedProducer;
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.ArcContainer;
+import io.quarkus.cache.Cache;
+import io.quarkus.cache.runtime.AbstractCache;
+import io.smallrye.mutiny.Uni;
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.impl.ContextInternal;
+
+/**
+ * This class is an internal Quarkus cache implementation using Infinispan Embedded.
+ * Do not use it explicitly from your Quarkus application.
+ */
+public class InfinispanCacheImpl extends AbstractCache implements Cache {
+
+    private final AdvancedCache cache;
+    private final InfinispanCacheInfo cacheInfo;
+    private final Map<Object, CompletableFuture> computationResults = new ConcurrentHashMap<>();
+    private final long lifespan;
+    private final long maxIdle;
+
+    public InfinispanCacheImpl(InfinispanCacheInfo cacheInfo, org.infinispan.Cache cache) {
+        this(cacheInfo, cache.getAdvancedCache());
+    }
+
+    public InfinispanCacheImpl(InfinispanCacheInfo cacheInfo, AdvancedCache cache) {
+        this.cacheInfo = cacheInfo;
+        this.cache = cache;
+        this.lifespan = cacheInfo.lifespan.map(l -> l.toMillis()).orElse(-1L);
+        this.maxIdle = cacheInfo.maxIdle.map(m -> m.toMillis()).orElse(-1L);
+    }
+
+    public InfinispanCacheImpl(InfinispanCacheInfo cacheInfo) {
+        this(cacheInfo, determineInfinispanClient(cacheInfo.name));
+    }
+
+    private static AdvancedCache determineInfinispanClient(String cacheName) {
+        ArcContainer container = Arc.container();
+        InfinispanEmbeddedProducer producer = container.instance(InfinispanEmbeddedProducer.class).get();
+        return producer.getAdvancedCache(cacheName);
+    }
+
+    @Override
+    public String getName() {
+        return Objects.requireNonNullElse(cacheInfo.name, "default-infinispan-cache");
+    }
+
+    @Override
+    public Object getDefaultKey() {
+        return "default-key";
+    }
+
+    private Object encodeNull(Object value) {
+        return value != null ? value : NullValue.NULL;
+    }
+
+    private <T> T decodeNull(Object value) {
+        return value != NullValue.NULL ? (T) value : null;
+    }
+
+    @Override
+    public <K, V> Uni<V> get(K key, Function<K, V> valueLoader) {
+        return Uni.createFrom()
+                .completionStage(() -> CompletionStages.handleAndCompose(cache.getAsync(key), (v1, ex1) -> {
```

Comment:
This can just be `thenHandle`, no reason for `handleAndCompse` from what I can see.